_posts/2018-09-26-Inter-service-communication.md

```diff
@@ -0,0 +1,216 @@
+
+{% highlight java %}
+
+   @Bean
+   AuthClient authClient() {
+        return Feign.builder()
+                .requestInterceptor(requestInterceptor())
+                .target(AuthClient.class, baseServerUrl);
+    }
+{% endhighlight %}
+
+### Client
+By default, the default HTTP client of Feign uses `HttpUrlConnection` to execute its HTTP requests.
+You can configure another client (`ApacheHttpClient`, `OkHttpClient`, ...) as follows:
+{% highlight java %}
+
+   @Bean
+   AuthClient authClient() {
+        return Feign.builder()
+                .client(new ApacheHttpClient())
+                .target(AuthClient.class, baseServerUrl);
+    }
+{% endhighlight %}
+
+#### OkHttpClient
+OkHttp is an HTTP client that’s efficient by default:
+
+* HTTP/2 support allows all requests to the same host to share a socket.
+* Connection pooling reduces request latency (if HTTP/2 isn’t available).
+* Transparent GZIP shrinks download sizes.
+* Response caching avoids the network completely for repeat requests.
+
+#### Mutual SSL with ApacheHttpClient
+The advantage of using `ApacheHttpClient` over the default client is that `ApacheHttpClient` sends more headers with the request, eg. `Content-Length`, which some servers expect.
+Aside, Apache brings support for mutual SSL, where we can safely store our key and trust-store into transit.
+To achieve this, you have to create a custom `ApacheHttpClient` and insert it into our Feign client. 
+
+
+### Retry mechanism
+When we want to build some resilience in our communication, we can setup a retry mechanism in our Feign client. 
+If the other service is unreachable, we will try again until it is healthy or until our configuration has been set. 
+When we want to use the retryer of Feign, we got three properties we can set.
+
+* Period: How long it takes before the retry is triggered
+
+* MaxPeriod: That's what the maximum is of how long it can take before a retry is triggered
+
+* MaxAttempts: How many retries may the client trigger before it fails
+
+Example:
+{% highlight java %}
+
+   @Bean
+   AuthClient authClient() {
+        return Feign.builder()
+                .retryer(new Retryer.Default(period, maxPeriod, maxAttempts))
+                .target(AuthClient.class, baseServerUrl);
+    }
+{% endhighlight %}
+
+### Encoder / Decoder
+Besides JSON encoders and decoders, you can also enable support for XML.
+If you ever have to integrate with SOAP third party API's, Feign supports it.
+There is a very detailed explanation on how to use it in the [documentation](https://github.com/OpenFeign/feign){:target="_blank"} of OpenFeign.
+
+# Security
+Security itself is very important in the structure of your platform, that's why we have to enable what mechanisms we can apply to the Feign client. 
+Implementations are give above in the examples. 
+
+## Basic
+>When you want to send basic credentials you can just add an interceptor for the Feign client and add the username and password.
+
+## JWT
```

Comment:
JWT isn't a type of authentication. I'd suggest to change this to "Bearer" or "Token" with OAuth2 and/or OIDC as methods. WDYT?